src/Dependencies/Collections/RoslynImmutableInterlocked.cs

```diff
@@ -0,0 +1,369 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.CodeAnalysis.Collections
+{
+    internal static class RoslynImmutableInterlocked
+    {
+        /// <summary>
+        /// Mutates a value in-place with optimistic locking transaction semantics via a specified transformation
+        /// function. The transformation is retried as many times as necessary to win the optimistic locking race.
+        /// </summary>
+        /// <typeparam name="TKey">The type of key stored by the dictionary.</typeparam>
+        /// <typeparam name="TValue">The type of value stored by the dictionary.</typeparam>
+        /// <param name="location">
+        /// The variable or field to be changed, which may be accessed by multiple threads.
+        /// </param>
+        /// <param name="transformer">
+        /// A function that mutates the value. This function should be side-effect free,
+        /// as it may run multiple times when races occur with other threads.</param>
+        /// <returns>
+        /// <see langword="true"/> if the location's value is changed by applying the result of the
+        /// <paramref name="transformer"/> function; otherwise, <see langword="false"/> if the location's value remained
+        /// the same because the last invocation of <paramref name="transformer"/> returned the existing value.
+        /// </returns>
+        public static bool Update<TKey, TValue>(ref ImmutableSegmentedDictionary<TKey, TValue> location, Func<ImmutableSegmentedDictionary<TKey, TValue>, ImmutableSegmentedDictionary<TKey, TValue>> transformer)
+            where TKey : notnull
+        {
+            if (transformer is null)
+                throw new ArgumentNullException(nameof(transformer));
+
+            bool successful;
+            var oldValue = ImmutableSegmentedDictionary<TKey, TValue>.PrivateInterlocked.VolatileRead(ref location);
+            do
+            {
+                var newValue = transformer(oldValue);
+                if (oldValue == newValue)
+                {
+                    // No change was actually required.
+                    return false;
+                }
+
+                var interlockedResult = InterlockedCompareExchange(ref location, newValue, oldValue);
+                successful = oldValue == interlockedResult;
```

Comment:
>oldValue == interlockedResult [](start = 29, length = 29)

If I'm reading this correctly, this is struct comparison, correct? Is there some reference-based comparison we can do to make it faster? Same for the oldValue == newValue above.

Reply:
`operator ==` is overloaded in exactly the same way it appears for `ImmutableArray<T>`. I can't use the same pattern that `ImmutableInterlocked` uses for `ImmutableArray<T>` because the underlying dictionary is declared `private`.